### PR TITLE
sql/parser: split into more separate packages

### PR DESCRIPTION
**Pull request by knz**
First commit from #19901 - will rebase when that merges.

tl;dr this patch separates items from the `sql/parser` package into
the following:

- new package `sem/builtins`: the built-in function *definitions*

- new package `sem/transforms`: `ContainsWindow`,
  `IsAggregateVisitor`, `ExprTransformContext` (not
  `NormalizeVisitor`, that's still entangled with `parser`)

- new package `sem/tests`: tests that need both the parser and the
  built-in function definitions, to break the dependency cycle.

Details follow.

The following `parser` definitions remain in the `parser` package for
now but are moved to their own files:

- `FunctionClass` and constants, `Builtin` and methods (`builtins.go` -> `builtin.go`)
- `AggregateFunc` (`aggregate_builtins.go` -> `aggregate_funs.go`)
- `ValueGenerator` (`generator_builtins.go` -> `generators.go`)
- `IndexedRow`, `WindowFrame`, `WindowFunc` (`window_builtins.go` -> `window_funs.go`)
- `intPow` (`builtins.go` -> `eval.go`)
- `pickFromTuple` (`builtins.go` -> `eval.go`)
- `mockNameTypes` (`type_check.go` -> `testutils.go`)
- `init`, (`builtins.go` -> `all_builtins.go`, then moved below)
- `collectBuiltins` (`window_builtins.go` -> `all_builtins.go`, then moved below)

- `TestCastToCollatedString` (`eval_test.go` -> `collatedstring_test.go`)
- `TestEvalComparisonExprCaching` (`eval_test.go` -> `compare_test.go`)
- `TestSimilarEscape`, `Benchmark*Like*` (`eval_test.go` -> `like_test.go`)
- `TestClusterTimestampConversion` (`eval_test.go` -> `timestamp_test.go`)

This patch then extracts code into the following separate modules:

- `canBeInArray` -> `IsValidArrayElementType` in `sql/sem/types`

- `ContainsWindowVisitor` -> `window.go` in `sql/sem/transform`
- `IsAggregateVisitor` -> `aggregates.go` in `sql/sem/transform`
- `expr_transform.go` -> `sql/sem/transform`

- `all_builtins.go`, `builtins.go` -> `sql/sem/builtins`
- `aggregate_builtins.go` -> `sql/sem/builtins`
- `aggregate_builtins_test.go` -> `sql/sem/builtins`
- `window_builtins.go` -> `sql/sem/builtins`
- `pg_builtins.go` -> `sql/sem/builtins`
- `EmptyDTable()` -> `sql/sem/builtins` (needs `arrayValueGenerator`, a built-in)

- `TestHelpFunctions` (`help_test.go` -> `sql/sem/builtins/help_test.go`)

The following tests are then in turn extracted into a new `sem/tests`
package, because they have a dependency on both `parser` and the new
`sem/builtins` package (they depend on the SQL built-in functions
being defined):

- `TestTypeCheck`, `TestTypeCheckError` -> (`type_check_test.go` -> `sql/sem/tests/type_check_test.go`)
- all tests from `format_test.go`
- all tests from `function_name_test.go`
- all tests from `normalize_test.go`
- all tests from `parse_test.go`
- all tests from `timeconv_test.go`
- remaining tests in `eval_test.go`

This in turn required the following API exports in `parser`:

- the fields of `Builtin`. This in turn removes the
  need for its public accessors.
- `AddWithOverflow`
- `AppendToMaybeNullArray`
- `ConcatArrays`
- `ErrDivByZero`
- `ErrZeroModulus`
- `EvalContext.StmtTimestamp`
- `EvalContext.TxnTimestamp`
- `FixedReturnType`, `UnknownReturnType`, `IdentityReturnType`
- `FunDefs`
- `IntPow`
- `MockNameTypes`
- `NewFunctionDefinition`
- `NormalizeVisitor`
- `PickFromTuple`
- `PrependToMaybeNullArray`
- `RegexpCacheKey` and its `Pattern` interface method.
- `ReturnTyper`
- `SecondsInDay`
- `TypeList`

The new following new items were added:

- `TimestampDifference`, to provide visibility into
  `parser.timestampMinusBinOp` from `sem/builtins` and clarify its
  intent.
- `FmtAlwaysGroupExprs` to provide access to the `alwaysParens` format
  flag from `sem/tests`.

The following was deleted, because unnecessary:

- `isNaN()` - inlined in caller, simplifies and reduces overhead

### CODE REVIEW

**Comment by knz** [comment]
@jordanlewis this is not quite ready to review yet. I'll add a 2nd commit that removes the alias stubs. However in the meantime can you confirm that the commit message is closer to what you'd like to see for #19901?

**Comment by knz** [comment]
hold on to reviewing this, I'm splitting it up in smaller patches to make it more easy.

**Comment by knz** [comment]
Superseded by #19941.
